sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java

```diff
@@ -690,47 +715,136 @@ public Write withQueue(String queue) {
-    public Write withPassword(String password) {
+    public Write<EventT> withPassword(String password) {
       checkArgument(password != null, "password can not be null");
       return builder().setPassword(password).build();
     }
 
+    /**
+     * Specify the JMS topic destination name where to send messages to dynamically. The {@link
+     * JmsIO.Write} acts as a publisher on the topic.
+     *
+     * <p>This method is exclusive with {@link JmsIO.Write#withQueue(String) and
+     *{@link JmsIO.Write#withTopic(String)}. The user has to specify a Serializable function
+     * that takes the Event Object as parameter, and returns the topîc name depending of the content
+     * of the event object.
+     *
+     * <p>For instance:
+     * <pre>{@code
+     * SerializableFunction<SomeEventObject, String> topicNameMapper =
+     *         (e ->
+     *             return the topic name ;
+     *
+     * }</pre>
+     *
+     * <pre>{@code
+     * .apply(JmsIO.write().withTopicNameMapper(topicNameNapper)
+     *
+     * }</pre>
+     *
+     * @param topicNameMapper The function returning the dynamic topic name.
+     * @return The corresponding {@link JmsIO.Write}.
+     */
+    public Write<EventT> withTopicNameMapper(SerializableFunction<EventT, String> topicNameMapper) {
+      checkArgument(topicNameMapper != null, "topicNameMapper can not be null");
+      return builder().setTopicNameMapper(topicNameMapper).build();
+    }
+
+    /**
+     * Map the Event object with a {@link javax.jms.Message}.
+     *
+     * <p>For instance:
+     *
+     * <pre>{@code
+     * SerializableMapper<VehicleEvent> valueMapper = (e, s) -> {
+     *
+     *       try {
+     *         TextMessage msg = s.createTextMessage();
+     *         msg.setText(Mapper.MAPPER.toJson(e));
+     *         return msg;
+     *       } catch (JMSException ex) {
+     *         throw new JmsIOException("Error!!", ex);
+     *       }
+     *     };
+     *
+     * }</pre>
+     *
+     * <pre>{@code
+     * .apply(JmsIO.write().withValueMapper(valueNapper)
+     *
+     * }</pre>
+     *
+     * @param valueMapper The function returning the {@link javax.jms.Message}
+     * @return The corresponding {@link JmsIO.Write}.
+     */
+    public Write<EventT> withValueMapper(SerializableMessageMapper<EventT> valueMapper) {
+      checkArgument(valueMapper != null, "valueMapper can not be null");
+      return builder().setValueMapper(valueMapper).build();
+    }
+
+    public Write<EventT> withCoder(Coder<EventT> coder) {
+      checkArgument(coder != null, "coder can not be null");
+      return builder().setCoder(coder).build();
+    }
+
     @Override
-    public PDone expand(PCollection<String> input) {
+    public WriteJmsResult<EventT> expand(PCollection<EventT> input) {
       checkArgument(getConnectionFactory() != null, "withConnectionFactory() is required");
+      checkArgument(getCoder() != null, "withCoder() is required");
```

Comment:
Why not use input.getCoder()?

This would allow you to get rid of withCoder().

Reply:
Yes, you're right